cpp/src/arrow/compute/exec/key_compare_avx2.cc

```diff
@@ -25,160 +25,545 @@ namespace compute {
+      reinterpret_cast<uint64_t*>(match_bytevector)[i] =
+          compare8_fn(rows_left, rows_right, i * unroll, irow_left, offset_right);
+
+      if (!use_selection) {
+        irow_left = _mm256_add_epi32(irow_left, _mm256_set1_epi32(8));
+      }
+    }
+    return num_rows_to_compare - (num_rows_to_compare % unroll);
   }
+}
 
-  uint32_t num_rows_processed = num_rows;
-  return num_rows_processed;
+template <bool use_selection>
+uint32_t KeyCompare::CompareBinaryColumnToRowImp_avx2(
+    uint32_t offset_within_row, uint32_t num_rows_to_compare,
+    const uint16_t* sel_left_maybe_null, const uint32_t* left_to_right_map,
+    KeyEncoder::KeyEncoderContext* ctx, const KeyEncoder::KeyColumnArray& col,
+    const KeyEncoder::KeyRowArray& rows, uint8_t* match_bytevector) {
+  uint32_t col_width = col.metadata().fixed_length;
+  if (col_width == 0) {
+    int bit_offset = col.bit_offset(1);
+    return CompareBinaryColumnToRowHelper_avx2<use_selection>(
+        offset_within_row, num_rows_to_compare, sel_left_maybe_null, left_to_right_map,
+        ctx, col, rows, match_bytevector,
+        [bit_offset](const uint8_t* left_base, const uint8_t* right_base,
+                     uint32_t irow_left_base, __m256i irow_left, __m256i offset_right) {
+          __m256i left;
+          if (use_selection) {
+            irow_left = _mm256_add_epi32(irow_left, _mm256_set1_epi32(bit_offset));
+            left = _mm256_i32gather_epi32((const int*)left_base,
+                                          _mm256_srli_epi32(irow_left, 3), 1);
+            left = _mm256_and_si256(
+                _mm256_set1_epi32(1),
+                _mm256_srlv_epi32(left,
+                                  _mm256_and_si256(irow_left, _mm256_set1_epi32(7))));
+            left = _mm256_mullo_epi32(left, _mm256_set1_epi32(0xff));
+          } else {
+            __m256i bits = _mm256_setr_epi32(1, 2, 4, 8, 16, 32, 64, 128);
+            uint32_t start_bit_index = irow_left_base + bit_offset;
+            uint8_t left_bits_8 =
+                (reinterpret_cast<const uint16_t*>(left_base + start_bit_index / 8)[0] >>
+                 (start_bit_index % 8)) &
+                0xff;
+            left = _mm256_cmpeq_epi32(
+                _mm256_and_si256(bits, _mm256_set1_epi8(left_bits_8)), bits);
+            left = _mm256_and_si256(left, _mm256_set1_epi32(0xff));
+          }
+          __m256i right = _mm256_i32gather_epi32((const int*)right_base, offset_right, 1);
+          right = _mm256_and_si256(right, _mm256_set1_epi32(0xff));
+          __m256i cmp = _mm256_cmpeq_epi32(left, right);
+          uint32_t result_lo =
+              _mm256_movemask_epi8(_mm256_cvtepi32_epi64(_mm256_castsi256_si128(cmp)));
+          uint32_t result_hi = _mm256_movemask_epi8(
+              _mm256_cvtepi32_epi64(_mm256_extracti128_si256(cmp, 1)));
+          return result_lo | (static_cast<uint64_t>(result_hi) << 32);
+        });
+  } else if (col_width == 1) {
```

Comment:
These blocks are very repetitive and it's difficult to see how they differ. As written the blocks look sufficiently different to make generic extraction difficult, which is unfortunate.

Reply:
I refactored that part a bit.

Reply:
That's an improvement, thanks!